pyodide-build/pyodide_build/mkpkg.py

```diff
@@ -151,10 +293,131 @@ def make_package(
+                    package_metadata["info"]["requires_dist"]
+                )
+                for req in reqs:
+                    m = req.marker
+                    extra_name = None
+                    if m != None:
+                        for m in m._markers:
+                            if str(m[0]) == "extra" and len(m) == 3:
+                                extra_name = str(m[2])
+                    dm.setdefault(extra_name, []).append(req)
+
+            return dm
+
+    class EnvironmentHelper(pkg_resources.Environment):
+        def __init__(self):
+            super().__init__(search_path=[packages_dir])
+
+        def scan(self, search_path=None):
+            pass
+
+        def make_dist(self, proj_name, meta_path):
+            yaml = YAML()
+            p_yaml = yaml.load(meta_path.read_bytes())
+            p_version = p_yaml["package"]["version"]
+            dist = YamlDistribution(
+                meta_path.parents[0], project_name=proj_name, version=p_version
+            )
+            return dist
+
+        def best_match(
+            self,
+            req,
+            working_set,
+            installer=None,
+            replace_conflicting=False,
+            from_install=False,
+        ):
+            proj_name = req.project_name
+            if os.path.isfile(packages_dir / proj_name / "meta.yaml"):
+                return self.make_dist(
+                    req.project_name, packages_dir / proj_name / "meta.yaml"
+                )
+            proj_name = proj_name.replace("-", "_")
+            if os.path.isfile(packages_dir / proj_name / "meta.yaml"):
+                return self.make_dist(
+                    req.project_name, packages_dir / proj_name / "meta.yaml"
+                )
+            # no package installed - try to install it
+            if from_install:
+                return None
+            parser = make_parser(argparse.ArgumentParser())
+            print("Installing dependency:", req.project_name, str(req.specs))
+            make_package(
+                packages_dir,
+                req.project_name,
+                str(req.specifier),
+                source_fmt=source_fmt,
+                make_dependencies=make_dependencies,
+                find_imports=find_imports,
+            )
+            return self.best_match(
+                req, working_set, installer, replace_conflicting, True
+            )
+
+    if extra:
+        our_extras = extra.split(",")
+    else:
+        our_extras = []
+    requires_packages = []
+    yaml_requires = []
+
+    package_dir = packages_dir / package
+
+    dist = YamlDistribution(package_dir, project_name=package, version=version)
+    requires_packages = dist.requires(extras=our_extras)
+
+    env = EnvironmentHelper()
+
+    if make_dependencies:
+        ws = pkg_resources.WorkingSet([])
+        ws.resolve(requires_packages, env=env, extras=our_extras)
+
+    for r in requires_packages:
```

Comment:
I think it's a bit problematic to try to re-compute dependencies, because it's a hard problem it's never going to be right exactly, and it feels a bit redundant to what we do in micropip. It's just a lot of code to maintain.

If we wanted to resolve dependencies, maybe, we could run
```
pip download <package-name>
```
in some temporary folder?
Which will do correct dependency resolution by pip, and then check the wheels it downloaded into the temp folder? Bonus point is that most of those would be wheels so they would contain the `top_level.txt` file with imports. Though one would still need to fetch their json metadata and in the case of a binary wheel take the .tar.gz.